qemu/tests/block_performance_test.py

```diff
@@ -1,111 +1,534 @@
 
-    if stg1_result > stg0_result * 1.05:
-        logger.warn("Unexpected result: stg0:%s stg1:%s", stg0_result,
-                    stg1_result)
-    else:
-        logger.info("Get result: stg0:%s stg1:%s", stg0_result, stg1_result)
+    def execute_operation(where, cmd):
+        # function
+        if cmd in locals_var:
+            logger.debug("Execute function %s", cmd)
+            return locals_var[cmd]()
+        # cmd line
+        if where == "host":
+            output = process.getoutput(cmd, timeout=cmd_timeout)
+        else:
+            output = session.cmd_output(cmd, timeout=cmd_timeout)
+        logger.debug("Execute %s cmd %s:output:\n%s", where, cmd, output)
+
+    logger = test.log
+    os_type = params["os_type"]
+    login_timeout = params.get_numeric("login_timeout", 360)
+    guest_operation = params.get("guest_operation")
+    fio_interval = params.get_numeric("fio_interval")
+    cmd_timeout = params.get_numeric("cmd_timeout", 1800)
+    run_times = params.get_numeric("run_times", 1)
+    fio_dir = params["fio_dir"]
+    test_results = {}
+    host_test_cmd = params["host_test_cmd"]
+    host_test_file = fio_dir + "/test.img"
+    check_host_iops_req = params.get_numeric("check_host_iops_req", 0)
+    guest_init_operation = params.get("guest_init_operation")
+    host_init_operation = params.get("host_init_operation")
+    guest_deinit_operation = params.get("guest_deinit_operation")
+    host_deinit_operation = params.get("host_deinit_operation")
+    sampling_rate = float(params.get("sampling_rate", "0.8"))
+    dispersion = float(params.get("dispersion", "0.1"))
+    error_threshold = float(params.get("error_threshold", "0.1"))
+    warn_threshold = float(params.get("warn_threshold", "0.05"))
+    boot_wait_time = params.get_numeric("boot_wait_time", 60)
+    select_disk_request = params.get("select_disk_request", "no")
+    select_disk_name = params.get("select_disk_name", "")
+    select_disk_minimum_size = params.get_numeric("select_disk_minimum_size",
+                                                  20)
+    vm = None
+
+    locals_var = locals()
+
+    if host_init_operation:
+        logger.debug("Execute host init :  %s" % host_init_operation)
+        execute_operation("host", host_init_operation)
+
+    auto_select_disk()
+
+    if check_host_iops_req > 0:
+        check_host_iops(host_test_file, check_host_iops_req)
+
+    if params["start_vm"] != 'yes':
+        params['start_vm'] = 'yes'
+        boot_images = params.get("compare_boot_images",
+                                 params.get("compare_images", " "))
+        params['images'] += boot_images + params.get("extra_images", " ")
+        logger.debug("Ready boot VM : %s", params['images'])
+        env_process.process(test, params, env, env_process.preprocess_image,
+                            env_process.preprocess_vm)
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
```

Comment:
FYI:
Based on Zhenchao's previous suggestion, try best to avoid modifying a cartesian param at runtime, we can use it in the following way.
cfg:
(In this way, we can remove line 500 - 503)
start_vm = no
images +=  ${compare_images} 
images += ${extra_images}
py:
vm_name = params["main_vm"]
env_process.preprocess_vm(test, params, env, vm_name)
vm = env.get_vm(vm_name)
# start_vm = no, not run vm.create in preprocess_vm, so need to add it
vm.create(params=params)
vm.verify_alive()

Reply:
The image files need to be built at this time. So it needs to be customized.

Reply:
updated in cfg file

Reply:
Sorry for the inconvenience. After our online discussion, I looked at the relevant code again and I suggest using the following method.
cfg:
not_preprocess = yes
py:
if params["not_preprocess"] != 'no':
    env_process.process(test, params, env, env_process.preprocess_image,
                                          env_process.preprocess_vm)
vm = env.get_vm(params["main_vm"])
vm.verify_alive()